src/inputfetcher.h

```diff
@@ -0,0 +1,197 @@
+#include <util/hasher.h>
+#include <util/threadnames.h>
+
+#include <atomic>
+#include <barrier>
+#include <cstdint>
+#include <stdexcept>
+#include <thread>
+#include <unordered_set>
+#include <utility>
+#include <vector>
+
+/**
+ * Helper for fetching inputs from the CoinsDB and CoinsTip and inserting them
+ * into the ephemeral cache used in ConnectBlock.
+ *
+ * It spawns a fixed set of worker threads that fetch Coins for each input
+ * in a block. The Coin is moved into the Input struct and then the status is
+ * atomically updated to READY. The main thread spin loops on the status field
+ * until it is READY and then inserts it into the temporary cache.
+ *
+ * Worker threads are synchronized with the main thread using a barrier, which
+ * is used at the beginning of fetching to start the workers and at the end to
+ * make sure all workers have exited the work loop.
+ */
+class InputFetcher
+{
+private:
+    //! The latest input being fetched. Workers atomically increment this when fetching.
+    alignas(64) std::atomic_size_t m_input_head{0};
+
+    //! The inputs of the block which is being fetched.
+    struct Input {
+        enum class Status : uint8_t {
+            WAITING, // The coin has not been fetched yet
+            READY, // The coin has been fetched and is ready to be inserted into the cache
+            FAILED, // The coin failed to be fetched
+            SKIPPED, // The coin is created and spent in the same block so cannot be fetched
+        };
+
+        //! Workers update this after setting the coin. The main thread spins on this until it is not WAITING.
+        std::atomic<Status> status{Status::WAITING};
+        //! The outpoint to fetch;
+        const COutPoint& outpoint;
+        //! The coin that workers will fetch and main thread will insert into cache.
+        Coin coin{};
+
+        Input(Input&& other) noexcept : outpoint{other.outpoint} {} // Only moved in setup for reallocation.
+        Input(const COutPoint& o LIFETIMEBOUND) noexcept : outpoint{o} {}
+    };
+    std::vector<Input> m_inputs{};
+
+    /**
+     * The set of txids of all txs in the block being fetched.
+     * Used to filter out inputs that are created and spent in the same block,
+     * since they will not be in the db or the cache.
+     */
+    std::unordered_set<Txid, SaltedTxidHasher> m_txids{};
+
+    //! DB coins view to fetch from.
+    const CCoinsView* m_db{nullptr};
+    //! The cache to fetch from.
+    const CCoinsViewCache* m_cache{nullptr};
+
+    std::vector<std::thread> m_worker_threads{};
+    std::barrier<> m_barrier;
+    bool m_request_stop{false};
+
+    void WorkLoop() noexcept
+    {
+        while (true) {
+            m_barrier.arrive_and_wait();
+            if (m_request_stop) [[unlikely]] {
+                return;
+            }
+            while (true) {
+                const size_t i{m_input_head.fetch_add(1, std::memory_order_relaxed)};
+                if (i >= m_inputs.size()) [[unlikely]] {
```

Comment:
I don't mind the `unlikely` parts but we're barely using them in the code, they have some weird side-effects when merged - which may not be the case here, so I'm fine wither way, if nothing else, it documents the usage